### PR TITLE
Add floating snackbar

### PR DESCRIPTION
**Pull request by alexander-schranz**
| Q | A
| --- | ---
| Bug fix? | no 
| New feature? |  yes
| BC breaks? | no 
| Deprecations? | no  
| Fixed tickets | fixes # 
| Related issues/PRs | # 
| License | MIT
| Documentation PR | sulu/sulu-docs# 

#### What's in this PR?

Add floating snackbar.

snackbar

#### Why?

To show snackbar messages for copy and pasting of blocks.

#### To Do

- [x] Implement Snackbar Container
- [x] Implement block messages

### CODE REVIEW

File: src/Sulu/Bundle/AdminBundle/Resources/translations/admin.en.json [hunk 1]
```diff
@@ -61,6 +61,8 @@
     "sulu_admin.list_search_placeholder": "Search…",
     "sulu_admin.error": "Error",
     "sulu_admin.warning": "Warning",
+    "sulu_admin.info": "Info",
+    "sulu_admin.success": "Success",
```
Comment (luca-rath) [inline src/Sulu/Bundle/AdminBundle/Resources/translations/admin.en.json:65]: Those are missing in `admin.de.json`, right?

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/BlockCollection.js [hunk 2]
```diff
@@ -24,6 +24,7 @@ type Props<T: string, U: {type: T}> = {|
     onChange: (value: Array<U>) => void,
     onSettingsClick?: (index: number) => void,
     onSortEnd?: (oldIndex: number, newIndex: number) => void,
+    onTriggerMessage?: (message: Message) => void,
```
Comment (luca-rath) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/BlockCollection.js]: I don't think `trigger` is the correct term for this use case, but I also can't find another one that fits better ...

**Review by luca-rath — CHANGES_REQUESTED** [review]

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/BlockCollection.js (hunk 2 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/BlockCollection.js]: I even didnt wanted to use "on" here but our eslint rule doesnt allow it. Alternative to trigger would be onDispatchMessage. What do you think?

**Comment by matthieu2607** [comment]
nice feature ;)

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockToolbar/BlockToolbar.js [hunk 3]
```diff
@@ -1,6 +1,6 @@
 // @flow
 import React from 'react';
-import classNames from 'classnames';
+import classnames from 'classnames';
```
Comment (niklasnatter) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockToolbar/BlockToolbar.js]: i think we imported this as `classNames` in all other componenets. not sure if there is a reason, but i think i would keep things consistent 🙈

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Snackbar/README.md [hunk 4]
```diff
@@ -1,13 +1,47 @@
...
+        <Snackbar behaviour="floating" message="Some warning is shown here" type="warning" />
+        
+        <Snackbar behaviour="floating" message="Some information is shown here" type="info" />
+        
+        <Snackbar behaviour="floating" message="Some not so bad went wrong" type="success" />
+    </div>
+</div>
+```
+
+There is also a behaviour `floating` option used for floating snackbar:
```
Comment (niklasnatter) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Snackbar/README.md]: this is already described this above right? 🤔

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/SnackbarContainer/snackbarContainer.scss [hunk 5]
```diff
@@ -0,0 +1,13 @@
+.container {
```
Comment (niklasnatter) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/SnackbarContainer/snackbarContainer.scss:1]: do you think we can reuse this somewhere? if this is specific to the `<Application>`, i think i would not extract a separate component

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/Application/Application.js [hunk 6]
```diff
@@ -208,6 +221,23 @@ class Application extends React.Component<Props>{
                             onClose={this.handleProfileOverlayClose}
                             open={this.openedProfileFormOverlay}
                         />
+                        {
+                            snackbarStore.messages.length
+                                ? <SnackbarContainer className={snackbarClass}>
```
Comment (niklasnatter) [inline src/Sulu/Bundle/AdminBundle/Resources/js/containers/Application/Application.js:226]: not sure if i would extract a `SnackbarContainer`

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/BlockCollection.js (hunk 2 above)
Comment (niklasnatter) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/BlockCollection.js]: i think would go with `onShowSnackbar` or `onDisplaySnackbar`

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Snackbar/Snackbar.js [hunk 7]
```diff
@@ -7,9 +7,11 @@ import {translate} from '../../utils/Translator';
 import Icon from '../Icon';
 import snackbarStyles from './snackbar.scss';
 
-export type SnackbarType = 'error' | 'warning';
+export type SnackbarType = 'error' | 'warning' | 'info' | 'success';
 
 type Props = {|
+    behaviour: 'static' | 'floating',
```
Comment (niklasnatter) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Snackbar/Snackbar.js]: this prop does not really change any behaviour, right? 🙈

Comment (niklasnatter) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/Snackbar/Snackbar.js]: it affects the style and the text, right? i think we have `skin` on a few other components. think that would describe it better

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/BlockCollection.js (hunk 2 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/BlockCollection.js]: I would go with `onDisplaySnackbar` then @luca-rath what do you think?

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/SnackbarContainer/snackbarContainer.scss (hunk 5 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/SnackbarContainer/snackbarContainer.scss:1]: yes inside an overlay that is way I created a component for it. While it is currented decided not to implement it, I would keep the component here.

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/Application/Application.js (hunk 6 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/AdminBundle/Resources/js/containers/Application/Application.js:226]: As written above the component could be used on other places.

File: src/Sulu/Bundle/AdminBundle/Resources/translations/admin.en.json (hunk 1 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/AdminBundle/Resources/translations/admin.en.json:65]: fixed

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/SnackbarContainer/snackbarContainer.scss (hunk 5 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/SnackbarContainer/snackbarContainer.scss:1]: This is general style and the application is handling when it is used in the application context.

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Snackbar/Snackbar.js (hunk 7 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/Snackbar/Snackbar.js]: Is fine for me, was also unsure about it.

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/BlockCollection.js [hunk 8]
```diff
@@ -147,37 +148,47 @@ class BlockCollection<T: string, U: {type: T}> extends React.Component<Props<T,
...
+        const elementsAfter = value.slice(insertionIndex);
+
+        // $FlowFixMe
+        onChange([...elementsBefore, ...newElements, ...elementsAfter]);
+        clipboard.set(BLOCKS_CLIPBOARD_KEY, undefined);
+
+        if (onDisplaySnackbar) {
+            onDisplaySnackbar({
+                type: 'info',
+                text: translate('sulu_admin.%count%_blocks_pasted', {count: newElements.length}),
```
Comment (niklasnatter) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/BlockCollection.js:189]: think i would call this `message` as it is also called `message` on the `Snackbar` and in the `Overlay`

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/types.js [hunk 9]
```diff
@@ -10,3 +10,9 @@ export type RenderBlockContentCallback<T: string, U: {type: T}>
     = (value: U, type: T, index: number, expanded: boolean) => Node;
 
 export type BlockMode = 'static' | 'sortable' | 'selectable';
+
+export type Message = {
```
Comment (niklasnatter) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/types.js:14]: maybe call this `SnackbarMessage`

File: src/Sulu/Bundle/AdminBundle/Resources/js/stores/snackbarStore/types.js [hunk 10]
```diff
@@ -0,0 +1,6 @@
+// @flow
+export type Message = {
+    icon?: string,
+    text: string,
```
Comment (niklasnatter) [inline src/Sulu/Bundle/AdminBundle/Resources/js/stores/snackbarStore/types.js:4]: think i would call this `message` as it is also called `message` on the `Snackbar` and in the `Overlay`

**Review by niklasnatter — COMMENTED** [review]
I am sorry, i did not see the `text`/`message` inconsistency before 😕

File: src/Sulu/Bundle/AdminBundle/Resources/js/stores/snackbarStore/types.js (hunk 10 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/AdminBundle/Resources/js/stores/snackbarStore/types.js:4]: find `message.message` strange.

Comment (alexander-schranz) [reply src/Sulu/Bundle/AdminBundle/Resources/js/stores/snackbarStore/types.js:4]: as discussed we will curently keep `message.text` maybe in future we should migrate all components using the `message` object instead have separate props for every thing. This effects the Snackbar, but also Dialog and maybe overlays which has a `snackbarMessage` and `snackbarType`.

**Pull request merged**